dns/src/record/utils.rs

```diff
@@ -0,0 +1,190 @@
+use std::fmt;
+use std::io::{self, Cursor, Read, Seek, SeekFrom};
+use std::ops::RangeInclusive;
+
+use crate::wire::*;
+
+/// A kinda hacky but alright way to avoid copying tons of data
+pub(crate) trait CursorExt {
+    /// Replace this when #[feature(cursor_remaining)] is stabilised
+    fn std_remaining_slice(&self) -> &[u8];
+
+    /// Convenience
+    fn truncated(&self, length: u64) -> Self;
+    fn with_truncated<T>(&mut self, length: u64, f: impl FnOnce(&mut Self, usize) -> T) -> T;
+}
+
+impl CursorExt for Cursor<&[u8]> {
+    fn std_remaining_slice(&self) -> &[u8] {
+        let inner = self.get_ref();
+        let len = self.position().min(inner.as_ref().len() as u64);
+        &inner[(len as usize)..]
+    }
+    fn truncated(&self, to_length: u64) -> Self {
+        let inner = self.get_ref();
+        let len = inner.len() as u64;
+        let start = self.position().min(len);
+        let end = (start + to_length).min(len);
```

Comment:
Oh. This should obviously just error if the remaining slice is not big enough to hold the requested length. And return a wire error